### PR TITLE
Store Property / Method list in ScriptInstanceData #647

### LINKED ISSUE #647: `ScriptInstance` implementation is currently unsound
In our script instance implementation, we pass a property list and method list to godot depending on slices returned from `get_property_list` and `get_method_list`. Godot will then later call [`free_property_list_func`](https://github.com/godot-rust/gdext/blob/fe5b02ba6abec0ba2d8fff612368a9c896872810/godot-core/src/engine/script_instance.rs#L560) and [`free_method_list_func`](https://github.com/godot-rust/gdext/blob/fe5b02ba6abec0ba2d8fff612368a9c896872810/godot-core/src/engine/script_instance.rs#L688) on each of these. In these methods we again call `get_property_list/get_method_list` to find the length of the array, and call `Vec::from_raw_parts(ptr, len, len)` on the pointer.

This is unsound, since there is nothing stopping the user from implementing `get_property_list` or `get_method_list` such that they return slices of different lengths each time. For instance we could do something like:
```rs
fn get_property_list(&self) -> &[PropertyInfo] {
  static COUNTER: AtomicUsize = AtomicUsize::new();
  let v = Box::leak(repeat_fn(|| PropertyInfo { .. /* initialize here */ }).take(COUNTER.fetch_add(1)).collect::<Vec<_>>());
  &*v
}
```
This would make the `free` callback always try to create a `Vec` with a too big length, thus UB.

# Solutions

In godot 4.3 we can use the new free-callbacks, which pass along the length of the array. So that means we have at least these options:

1. Deprecate and eventually remove `ScriptInstance` support from all versions prior to 4.3, use new callbacks in 4.3
2. Make a sound implementation in versions prior to 4.3, use the new callbacks in 4.3
3. Make a sound implementation that works in all versions

I think option 1 is best. As making a sound implementation is tricky (you can use a sentinel value for the length, or some kind of header struct trick). And in 4.3 we wont need anything fancy for the length since the length is simply passed to us by Godot.

### PR DESCRIPTION
**Pull request by TitanNano**
Instead of accepting a reference to a slice of `PropertyInfo`s or `MethodInfo`s from the trait and then hoping that the length didn't change when freeing the list, we now store the lists inside the `ScriptInstanceData`. 

When the engine requests to free the list pointer, we can use the list to recover the length and later drop the rust data as well. 

Fixes #647 _[edit Bromeon]_
cc @lilizoey

### CODE REVIEW

**Comment by GodotRust** [comment]
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-650

File: godot-core/src/engine/script_instance.rs [hunk 1]
```diff
@@ -382,10 +388,10 @@ mod script_instance_info {
...
     /// - The returned `*const T` is guaranteed to point to a list that has an equal length and capacity.
-    fn transfer_ptr_list_to_godot<T>(ptr_list: Vec<T>, list_length: &mut u32) -> *const T {
+    fn transfer_ptr_list_to_godot<T>(ptr_list: Box<[T]>, list_length: &mut u32) -> *const T {
         *list_length = ptr_list.len() as u32;
 
-        let ptr = Box::into_raw(ptr_list.into_boxed_slice());
+        let ptr = Box::into_raw(ptr_list);
 
         // SAFETY: `ptr` was just created in the line above and should be safe to dereference.
         unsafe { (*ptr).as_ptr() }
```
Comment (lilizoey) [inline godot-core/src/engine/script_instance.rs:397]: I've actually just noticed that this introduces a subtle kind of UB:

Here you call `(*ptr).as_ptr()`. `as_ptr()` takes `&self` as an argument. This means that the returned pointer is derived from an immutable reference. Thus this pointer and any derived pointers cannot mutably access this box-slice.

However later you end up using this pointer to call `Vec::from_raw_parts`. This requires unique access to the boxed slice. But this pointer does not have permission to uniquely access that slice. Thus this is UB.

This can be fixed by changing this call to `(*ptr).as_mut_ptr()`. Ideally you should change the return type here to `*mut T` to reflect this change, as well as changing the corresponding argument `transfer_ptr_list_from_godot` type signature to reflect that as well. But that isn't safety critical.

**Review by lilizoey — COMMENTED** [review]
The current implementation has some cases of UB that should be relatively easy to fix.

However on a more general point, by changing from returning slice references to an owned value, you've had to introduce the whole mutex + hashmap. is there a reason you chose to do this over just using Godot 4.3's new property list implementation which passes the length of the list to `free_property_list`? does this need to be available in a version of godot < 4.3?

**Comment by TitanNano** [comment]
>  is there a reason you chose to do this over just using Godot 4.3's new property list implementation which passes the length of the list to free_property_list? does this need to be available in a version of godot < 4.3?

@lilizoey I have a personal interest in keeping the compatibility with 4.2 as I am currently using it with 4.2 and 4.3 has not been released yet. 

Regardless of that, the current implementation works, beside this issue, with 4.2, so I think it's quite heavy-handed to remove support for 4.2 entirely. 

Additionally, the trait currently ensures that the `&[PropertyInfo]` outlives `&self` but it doesn't stop the slice from being mutated once the trait function returned. What happens when a `Vec` gets extended beyond its capacity and the buffer is reallocated? I assume it will have similar issues as what you pointed out with the `Box`. Therefore, I believe it's much saver for the unsafe code to hold ownership (and control) of the data it is handing raw pointers out to.

File: godot-core/src/engine/script_instance.rs [hunk 2]
```diff
@@ -505,23 +511,32 @@ mod script_instance_info {
...
         // SAFETY: list_length has to be a valid pointer to a u32.
         let list_length = unsafe { &mut *r_count };
+        let return_pointer = transfer_ptr_list_to_godot(property_sys_list, list_length);
+
+        let instance = instance_data_as_script_instance::<T>(p_instance);
+        instance
+            .property_list
+            .lock()
+            .expect("Mutex should not be poisoned")
+            .insert(return_pointer, property_list);
```
[the remark this answers is not in the document]
Comment (TitanNano) [reply godot-core/src/engine/script_instance.rs:537]: What about using an `Arc` instead of a Box? It would allow the implementor to reuse an existing slice if they wish to. I have never tried this and don't know how `Arc`s and raw pointers interact, but I assume you might know.

**Comment by Bromeon** [comment]
> @lilizoey I have a personal interest in keeping the compatibility with 4.2 as I am currently using it with 4.2 and 4.3 has not been released yet.

We need to balance this need with the code complexity in gdext, and 4.3 is expected to be released in April. It would be quite a bit of extra effort for us to test and maintain two separate implementations across this version boundary, especially considering that not many people seem to use `ScriptInstance` at the moment.

Is using a 4.3-dev version not an option for you?

---

> Regardless of that, the current implementation works, beside this issue, with 4.2, so I think it's quite heavy-handed to remove support for 4.2 entirely.

We would eventually still need to modernize it to make use of the improvements in 4.3, so the question is how we would want to do that. `#[cfg]` based conditional compilation? Or drop 4.2 support once 4.3 is released?

**Comment by TitanNano** [comment]
> Is using a 4.3-dev version not an option for you?

I would prefer not to rely on a development version of the engine, but of course, if it is the best way forward, it is an acceptable compromise.

> We would eventually still need to modernize it to make use of the improvements in 4.3, [...]

If the assumptions of my third paragraph are correct, we would have to keep most of this patch for 4.3 as well. Regrading 
> #[cfg] based conditional compilation? Or drop 4.2 support once 4.3 is released?

I would say whatever is more practical. What is the general idea for backwards compatibility with older engine versions?

**Comment by lilizoey** [comment]
> Additionally, the trait currently ensures that the `&[PropertyInfo]` outlives `&self` but it doesn't stop the slice from being mutated once the trait function returned. What happens when a `Vec` gets extended beyond its capacity and the buffer is reallocated? I assume it will have similar issues as what you pointed out with the `Box`. Therefore, I believe it's much saver for the unsafe code to hold ownership (and control) of the data it is handing raw pointers out to.

Doing so would require the user to use interior mutability to modify the vec while also handing out a reference to the contents of that vec. i'm not entirely sure what mechanism could be used to do that? if you could then im pretty sure you could have UB even in safe rust, since you could then do this:
```rs
// imagine some type that stores a vec it can both mutate and hand out a slice-reference to. 
impl SomeType {
  fn mutating_getter(&self) -> &[i32] {
    self.mutate_vec();
    self.get_vec_ref()
  }
}

fn main() {
  let foo = SomeType::new();
  let slice_1 = foo.mutating_getter();
  let slice_2 = foo.mutating_getter(); // this might then invalidate slice_1?
  println!("{slice_1:?}"); // this might be UB then 
}
```

**Comment by lilizoey** [comment]
Ok so i think there is a way. since we dont track the fact that there's a shared reference to the instance when execution is handed over to godot, that means that gd-cell potentially doesn't know that it can't hand out mutable references. *this* would let the vec be mutated and the slice may get invalidated.

So to solve this we would either need to:
1. store the shared ref guard gd-cell hands out, this likely needs mutex + hashmap too
2. add some way to "fake" a shared reference with gd-cell, for instance a pair of methods `increment_shared_ref` and an unsafe `decrement_shared_ref`.

additionally there's this solution.

finally i suspect it may be possible to avoid owning any values, and just temporarily leaking the strings in propertyinfo before reclaiming them in the free method. this might require a bit more work though so i dont think it's something that we should do in this PR. but it could work better as a long term solution.

**Comment by TitanNano** [comment]
@lilizoey i was thinking of something along the lines of:
```
-> engine calls get_property_list(...)
-> engine calls call("method_that_can_mutate_the_property_list") in an other thread
-> engine accesses a pointer in the property list
```
-----
I think solution 1. Is not really viable as we would have to make the trait receive and return the refguard which is something we can't do as GdCell types are to remain private. This was also my first approach but it felt too cumbersome for the implementor.

-----
How should we move forward for now?

**Comment by lilizoey** [comment]
I think it's fine to keep this solution as is for now.  

Maybe in the future we can look into one of the other solutions as i think they need some more infrastructure work. Since i think i agree that solution 1 isn't going to be the best and all the other solutions require figuring out some other things in the backend, either in gd-cell or propertyinfo.

File: godot-core/src/engine/script_instance.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lilizoey) [reply godot-core/src/engine/script_instance.rs:537]: I'm not entirely sure, but i dont think they have the same issue as box. `Box` is rather special in rust so has a bunch of extra rules that `Arc` shouldn't have. I think it'd be easier to use `Vec` though. 

If you have a particular reason it'd be better to use `Arc` than `Vec` then feel free to. But barring a use-case i think `Vec` is preferable since it's gonna be more familiar to more people.

[the remark this answers is not in the document]
Comment (lilizoey) [reply godot-core/src/engine/script_instance.rs:537]: Miri seems happy with `Arc`:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=21b9778a61c05f69b90f6de2a841200f

[the remark this answers is not in the document]
Comment (TitanNano) [reply godot-core/src/engine/script_instance.rs:537]: Use case would be to be able to reuse the same slice multiple times instead of returning a new one every call.

[the remark this answers is not in the document]
Comment (TitanNano) [reply godot-core/src/engine/script_instance.rs:537]: Using an Arc unfortunately causes a Clippy warning because `PropertyInfo` / `MethodInfo` is neither `Send` nor `Sync`. I think for the use case here that is not actually an issue, but this warning is raised at every `Arc::new([...])` so also in user code. I'm sticking with `Vec<PropertyInfo>`.

[the remark this answers is not in the document]
Comment (Bromeon) [reply godot-core/src/engine/script_instance.rs:537]: What about `Rc`?

I also think there's value in familiarity, and `Vec` is easier to use. Question is whether these functions are invoked so often that it could cause a problem...

File: godot-core/src/engine/script_instance.rs (hunk 1 above)
Comment (TitanNano) [reply godot-core/src/engine/script_instance.rs]: Fixed this.

File: godot-core/src/engine/script_instance.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TitanNano) [reply godot-core/src/engine/script_instance.rs:537]: @Bromeon the `Rc` could be cloned in `get_property_list_func` and will be dropped in `free_property_list_func` which can be called by the engine on multiple threads, so the reference count decrement would not be thread safe.

File: itest/rust/src/builtin_tests/script/script_instance_tests.rs [hunk 3]
```diff
@@ -38,7 +38,9 @@ impl IScriptExtension for TestScript {
 struct TestScriptInstance {
     /// A field to store the value of the `script_property_b` during tests.
     script_property_b: bool,
+    #[allow(clippy::arc_with_non_send_sync)]
     prop_list: Vec<PropertyInfo>,
+    #[allow(clippy::arc_with_non_send_sync)]
     method_list: Vec<MethodInfo>,
```
Comment (lilizoey) [inline itest/rust/src/builtin_tests/script/script_instance_tests.rs:44]:
```suggestion
    prop_list: Vec<PropertyInfo>,
    method_list: Vec<MethodInfo>,
```

These `allow`s shouldn't be needed here right?

**Review by lilizoey — COMMENTED** [review]
I think this looks good! Apart from that you left in some `allow`s by accident?

File: itest/rust/src/builtin_tests/script/script_instance_tests.rs (hunk 3 above)
Comment (TitanNano) [reply itest/rust/src/builtin_tests/script/script_instance_tests.rs]: thanks.

**Review by lilizoey — APPROVED** [review]
Looks good, thanks!

The CI error seems unrelated to this PR, so im going to just approve it for now. I'm not very familiar with how that part of the CI works so i think that's something you need to fix @Bromeon? I tried rerunning the job once but it still failed. Feel free to merge this when it works.

**Comment by Bromeon** [comment]
Let's see if the fix in #657 works...
